                        EDA on the Spotify dataset, especially in Indonesia

This is a personal project created to find out about music in Indonesia,
I'm just curious what kind of songs are played the most in Indonesia based on genre, etc. 

in this part i try to :
   - explore several features such as snapshot_date, daily_rank, popularity, is_explicit, danceability, energy, acousticness, valence, songs(name), artists
   
   - Change column data type
   - change column name
   - take some information from descriptive statistics
   - Check for duplicate data and null values
   - Number of songs containing explicit words
   - check the daily_rank column
   
   - explore songs that are at daily_rank = 1
   - Explore songs that are in the popularity 100
   - correlation between the features
   - Genre: danceability, energy and valence while in Indonesia

# LOAD THE DATASET

In [14]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

pd.set_option('display.max_columns', None)

In [15]:
df_spotify = pd.read_csv('C:/Users/ASUS/Downloads/top_spotify_songs.csv')
df_spotify

spotify_id                                 name  \
0       2KslE17cAJNHTsI2MI0jb2                 Standing Next to You   
1       2OzhQlSqBEmt7hmkYxfT6m        Fortnight (feat. Post Malone)   
2       2HRgqmZQC0MC7GeNuDIXHN  Seven (feat. Latto) (Explicit Ver.)   
3       2xOhv7XudrBDtkID1jwsFE                              Từng Là   
4       3qhYidu0cemx1v9PgTtpS5               Chúng Ta Của Tương Lai   
...                        ...                                  ...   
676948  0AYt6NMyyLd0rLuvr0UkMH            Slime You Out (feat. SZA)   
676949  2Gk6fi0dqt91NKvlzGsmm7    SAY MY GRACE (feat. Travis Scott)   
676950  26b3oVLrRUaaybJulow9kz                               People   
676951  5ydjxBSUIDn26MFzU3asP4                           Rainy Days   
676952  59NraMJsLaMCVtwXTSia8i                                Prada   

                            artists  daily_rank  daily_movement  \
0                         Jung Kook           1               0   
1         Taylor Swift, Post Malone           2               0   
2                  Jung Kook, Latto           3               1   
3                      Vũ Cát Tường           4              -1   
4                     Sơn Tùng M-TP           5               0   
...                             ...         ...             ...   
676948                   Drake, SZA          46               4   
676949         Offset, Travis Scott          47               3   
676950                     Libianca          48               2   
676951                            V          49               1   
676952  cassö, RAYE, D-Block Europe          50               0   

        weekly_movement country snapshot_date  popularity  is_explicit  \
0                     0      VN    2024-04-23          92        False   
1                    48      VN    2024-04-23          88        False   
2                    -1      VN    2024-04-23          87         True   
3                     0      VN    2024-04-23          71        False   
4                    -2      VN    2024-04-23          73        False   
...                 ...     ...           ...         ...          ...   
676948                0      AE    2023-10-18          84         True   
676949                0      AE    2023-10-18          80         True   
676950                0      AE    2023-10-18          88        False   
676951                0      AE    2023-10-18          88        False   
676952                0      AE    2023-10-18          94         True   

        duration_ms                     album_name album_release_date  \
0            206019                         GOLDEN         2023-11-03   
1            228965  THE TORTURED POETS DEPARTMENT         2024-04-18   
2            183550                         GOLDEN         2023-11-03   
3            252500                        Từng Là         2024-03-01   
4            249871         Chúng Ta Của Tương Lai         2024-03-08   
...             ...                            ...                ...   
676948       310490               For All The Dogs         2023-10-06   
676949       173253                     SET IT OFF         2023-10-13   
676950       184791                         People         2022-12-06   
676951       179560                        Layover         2023-08-11   
676952       132359                          Prada         2023-08-11   

        danceability  energy  key  loudness  mode  speechiness  acousticness  \
0              0.711   0.809    2    -4.389     0       0.0955        0.0447   
1              0.675   0.397   11   -10.895     1       0.0245        0.4990   
2              0.790   0.831   11    -4.185     1       0.0440        0.3120   
3              0.808   0.414    5   -10.950     1       0.0380        0.8640   
4              0.694   0.556    0    -7.097     1       0.0805        0.7870   
...              ...     ...  ...       ...   ...          ...           ...   
676948         0.483   0.408    5   

In [17]:
df_id = df_spotify[df_spotify['country']=='ID']
df_id

spotify_id                                      name  \
1200    6EIMUjQ7Q8Zr2VtIUik4He                                drunk text   
1201    51ZQ1vr10ffzbwIjDCwqm4  we can't be friends (wait for your love)   
1202    3qhlB30KknSejmIvZZLjOD                          End of Beginning   
1203    6XKEElbr3cWmnUw6a5HRBV                               Mati-Matian   
1204    6L1jkVbb331DNvK89C3L8u                              Masih Hatiku   
...                        ...                                       ...   
675448  3hUxzQpSfdDqwM3ZTFQY0K                                    august   
675449  1USA6DXJ1qsucoKnDu4vnu                            Tak Kan Hilang   
675450  0DVCvSb8PhHchIP23JOP9q                             Tertawan Hati   
675451  5IhULiOyG58jL20qWkztEg                                     Dumes   
675452  5LrN7yUQAzvthd4QujgPFr                              Here With Me   

                            artists  daily_rank  daily_movement  \
1200                   Henry Moodie           1               0   
1201                  Ariana Grande           2               1   
1202                            Djo           3              -1   
1203                       Mahalini           4               0   
1204    Arsy Widianto, Tiara Andini           5               2   
...                             ...         ...             ...   
675448                 Taylor Swift          46               4   
675449                  Budi Doremi          47               3   
675450                      Awdella          48               2   
675451         OMWAWES, Guyon Waton          49               1   
675452                         d4vd          50               0   

        weekly_movement country snapshot_date  popularity  is_explicit  \
1200                  0      ID    2024-04-23          86        False   
1201                  1      ID    2024-04-23          96        False   
1202                 -1      ID    2024-04-23          99        False   
1203                  1      ID    2024-04-23          78        False   
1204                  2      ID    2024-04-23          79        False   
...                 ...     ...           ...         ...          ...   
675448                0      ID    2023-10-18          92        False   
675449                0      ID    2023-10-18          79        False   
675450                0      ID    2023-10-18          75        False   
675451                0      ID    2023-10-18          75        False   
675452                0      ID    2023-10-18          90        False   

        duration_ms                  album_name album_release_date  \
1200         187263  in all of my lonely nights         2024-01-26   
1201         228639            eternal sunshine         2024-03-08   
1202         159245                      DECIDE         2022-09-16   
1203         244137                 Mati-Matian         2024-03-29   
1204         262341                Masih Hatiku         2024-02-22   
...             ...                         ...                ...   
675448       261922                    folklore         2020-07-24   
675449       241714              Tak Kan Hilang         2022-02-25   
675450       237863               Tertawan Hati         2022-07-01   
675451       266666                       Dumes         2023-06-30   
675452       242484            Petals to Thorns         2023-05-26   

        danceability  energy  key  loudness  mode  speechiness  acousticness  \
1200           0.614   0.493    0    -5.356     1       0.0314        0.0676   
1201           0.645   0.663    5    -8.305     1       0.0447        0.0657   
1202           0.689   0.454    2    -7.643     1       0.0584        0.0351   
1203           0.440   0.389    5    -6.412     1       0.0368        0.7950   
1204           0.596   0.466    9    -8.341     1       0.0309        0.8170   
...              ...     ...  ...       ...   ...          ...           ...   
675448      

## Change the column data type
column type snapshot_date and album_release_date should be type datetime64[ns]

In [18]:
df_id.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9356 entries, 1200 to 675452
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          9356 non-null   object 
 1   name                9356 non-null   object 
 2   artists             9356 non-null   object 
 3   daily_rank          9356 non-null   int64  
 4   daily_movement      9356 non-null   int64  
 5   weekly_movement     9356 non-null   int64  
 6   country             9356 non-null   object 
 7   snapshot_date       9356 non-null   object 
 8   popularity          9356 non-null   int64  
 9   is_explicit         9356 non-null   bool   
 10  duration_ms         9356 non-null   int64  
 11  album_name          9356 non-null   object 
 12  album_release_date  9356 non-null   object 
 13  danceability        9356 non-null   float64
 14  energy              9356 non-null   float64
 15  key                 9356 non-null   int64  
 16  l

In [19]:
print('column types snapshot_date and album_release_date are now of type datetime64[ns]\n')

df_id.loc[:,'snapshot_date'] = pd.to_datetime(df_id.loc[:,'snapshot_date'])
df_id.loc[:,'album_release_date'] = pd.to_datetime(df_id.loc[:,'album_release_date'])
df_id.info()

column types snapshot_date and album_release_date are now of type datetime64[ns]



C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


<class 'pandas.core.frame.DataFrame'>
Int64Index: 9356 entries, 1200 to 675452
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   spotify_id          9356 non-null   object        
 1   name                9356 non-null   object        
 2   artists             9356 non-null   object        
 3   daily_rank          9356 non-null   int64         
 4   daily_movement      9356 non-null   int64         
 5   weekly_movement     9356 non-null   int64         
 6   country             9356 non-null   object        
 7   snapshot_date       9356 non-null   datetime64[ns]
 8   popularity          9356 non-null   int64         
 9   is_explicit         9356 non-null   bool          
 10  duration_ms         9356 non-null   int64         
 11  album_name          9356 non-null   object        
 12  album_release_date  9356 non-null   datetime64[ns]
 13  danceability        9356 non-null   float64

# Change column name
The Name column is a column that contains a collection of songs being played. To make it easier to recognize this column, the column has been renamed the Songs column.

In [20]:
df_id.rename(columns={'name' : 'Songs'}, inplace=True)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\frame.py:4441: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


# Some information from descriptive statistics
- Data collection starts from 2023-10-18 to 2024-04-23
- The number of days calculated from 2023-10-18 to 2024-04-23 is 187 days
- The song Duka by Nadin Amizah appears the most in the dataset
- There are 106 songs from 85 artists
- in the data set the mean value of each genre, the rotated average:
    - Danceability = 0.528116	
    - Energy = 0.512441
    - Valnece = 0.385457
    - Acousticness = 0.516563

In [21]:
df_id.describe(include='all')

<ipython-input-21-49f1477aa211>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df_id.describe(include='all')
<ipython-input-21-49f1477aa211>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df_id.describe(include='all')


spotify_id  \
count                     9356   
unique                     117   
top     3e1rs346dsDDwpqTRGlRZR   
freq                       187   
first                      NaN   
last                       NaN   
mean                       NaN   
std                        NaN   
min                        NaN   
25%                        NaN   
50%                        NaN   
75%                        NaN   
max                        NaN   

                                                    Songs       artists  \
count                                                9356          9356   
unique                                                106            85   
top     Rewrite The Stars (with James Arthur & Anne-Ma...  Nadin Amizah   
freq                                                  187           413   
first                                                 NaN           NaN   
last                                                  NaN           NaN   
mean                                                  NaN           NaN   
std                                                   NaN           NaN   
min                                                   NaN           NaN   
25%                                                   NaN           NaN   
50%                                                   NaN           NaN   
75%                                                   NaN           NaN   
max                                                   NaN           NaN   

         daily_rank  daily_movement  weekly_movement country  \
count   9356.000000     9356.000000      9356.000000    9356   
unique          NaN             NaN              NaN       1   
top             NaN             NaN              NaN      ID   
freq            NaN             NaN              NaN    9356   
first           NaN             NaN              NaN     NaN   
last            NaN             NaN              NaN     NaN   
mean      25.486853        0.485036         1.617251     NaN   
std       14.430872        4.960376         9.383215     NaN   
min        1.000000      -29.000000       -32.000000     NaN   
25%       13.000000       -1.000000        -3.000000     NaN   
50%       25.000000        0.000000         0.000000     NaN   
75%       38.000000        1.000000         3.000000     NaN   
max       50.000000       49.000000        49.000000     NaN   

              snapshot_date   popularity is_explicit    duration_ms  \
count                  9356  9356.000000        9356    9356.000000   
unique                  187          NaN           2            NaN   
top     2024-03-18 00:00:00          NaN       False            NaN   
freq                     52          NaN        9084            NaN   
first   2023-10-18 00:00:00          NaN         NaN            NaN   
last    2024-04-23 00:00:00          NaN         NaN            NaN   
mean                    NaN    81.524583         NaN  236394.865648   
std                     NaN     7.168964         NaN   44820.806317   
min                     NaN     0.000000         NaN   92400.000000   
25%                     NaN    78.000000         NaN  209647.000000   
50%                     NaN    80.000000         NaN  232440.000000   
75%                     NaN    84.000000         NaN  268422.000000   
max                     NaN   100.000000         NaN  325147.000000   

                    album_name   album_release_date  danceability  \
count                     9356                 9356   9356.000000   
unique                     102                   99           NaN   
top     Menari Dengan Bayangan  2019-11-29 00:00:00           NaN   
freq                       308                  308           NaN   
first                      NaN  1983-03-03 00:00:00           NaN   
last                       NaN  2024-04-18 00:00:00           NaN   
mean                       NaN                  NaN      0.528116   
std                        NaN                  

## Check for duplicate data and null values

In [22]:
if df_id.duplicated().values.any():
    print("The dataset has duplicate entries.")
else:
    print('No duplicate entries in the dataframe.')

No duplicate entries in the dataframe.


In [23]:
[features for features in df_id.columns if df_id[features].isnull().sum()>0]

[]

no column is null

## There are 2 days/dates that were not recorded during data scraping, namely:
 - 2024-03-17
 - 2024-03-15

In [24]:
# there are several dates that are not included in the snapshot_date column
df_id['snapshot_date'].unique()

array(['2024-04-23T00:00:00.000000000', '2024-04-22T00:00:00.000000000',
       '2024-04-21T00:00:00.000000000', '2024-04-20T00:00:00.000000000',
       '2024-04-19T00:00:00.000000000', '2024-04-18T00:00:00.000000000',
       '2024-04-17T00:00:00.000000000', '2024-04-16T00:00:00.000000000',
       '2024-04-15T00:00:00.000000000', '2024-04-14T00:00:00.000000000',
       '2024-04-13T00:00:00.000000000', '2024-04-12T00:00:00.000000000',
       '2024-04-11T00:00:00.000000000', '2024-04-10T00:00:00.000000000',
       '2024-04-09T00:00:00.000000000', '2024-04-08T00:00:00.000000000',
       '2024-04-07T00:00:00.000000000', '2024-04-06T00:00:00.000000000',
       '2024-04-05T00:00:00.000000000', '2024-04-04T00:00:00.000000000',
       '2024-04-03T00:00:00.000000000', '2024-04-02T00:00:00.000000000',
       '2024-04-01T00:00:00.000000000', '2024-03-31T00:00:00.000000000',
       '2024-03-30T00:00:00.000000000', '2024-03-29T00:00:00.000000000',
       '2024-03-28T00:00:00.000000000', '2024-03-27

# Number of songs containing explicit words
of all the songs played,
a total of 9084 songs containing the words explicit and 
272 which does not contain the word explicit

In [26]:
df_id.is_explicit.value_counts()

False    9084
True      272
Name: is_explicit, dtype: int64

## check the daily_rank column  
The purpose of this check is to ensure that even though the daily_rank column does not have a null value, there is a possibility that the data filled in is not appropriate, such as there being the same ranking on the same day.

all data in the daily_rank column is filled in properly, namely 1-50 per day

In [27]:
# look for numbers that are not in the daily_rank column
# Check rankings that may be missing or non-existent
for snapshot_date in df_id['snapshot_date'].unique():
    daily = [1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]
    
    aa = df_id[df_id.snapshot_date==pd.Timestamp(snapshot_date)]

    diff = set(aa.daily_rank.to_list()).difference(set(daily))
    print(diff)

set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set()
set(

In [28]:
df_id.columns

Index(['spotify_id', 'Songs', 'artists', 'daily_rank', 'daily_movement',
       'weekly_movement', 'country', 'snapshot_date', 'popularity',
       'is_explicit', 'duration_ms', 'album_name', 'album_release_date',
       'danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness',
       'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo',
       'time_signature'],
      dtype='object')

# explore songs that are at daily_rank 1

songs that get daily ranking 1:

    penjaga hati for     145 days
	drunk text for        25 days
	End of Beginning for  14 days
	Tak Dianggap for       3 days
    
    
penjaga hati song appeared with the highest daily ranking of 1, namely for 145 days

the song penjaga hati was released on 06-21-2023 but can still get daily ranking 1 in 2024

sequentially starting from the date the data was taken on 10-18-2023 to 03-13-2024 except on 03-05-2024 and 03-06-2024 which did not get a daily_rank of 1

In [ ]:
daily_rank_1= df_id[df_id['daily_rank'] == 1]

songs_daily_rank_1 = df_id.groupby(['snapshot_date','Songs','album_release_date', daily_rank_1['daily_rank']]).size().reset_index().rename(columns={0:'jumlah'})
# songs_daily_rank_1.sort_values(by ='snapshot_date', ascending=False).head(26)



In [ ]:
groupby_songs_daily_rank_1 =songs_daily_rank_1.groupby('Songs').size().reset_index().rename(columns={0:'Amount'}).sort_values('Amount',ascending=False)
display(groupby_songs_daily_rank_1)

def lagu_daily_rank_1(song):
    music = daily_rank_1[(daily_rank_1['Songs'] == song)]
    print('album_release_date',music.album_release_date.head(1),'\n')
    print(music.snapshot_date.unique())
    
    
lagu_daily_rank_1('penjaga hati')

.




THE TENDENCY OF EACH RANKING 1 SONG IS TOWARDS A PARTICULAR GENRE
  
- 'penjaga hati'  song tend to be in the acoustic genre

- 'Drunk Text'  song tend to be acoustic and danceable

- 'End of Beginning'  song tend to fall into the valence and danceability genres

- 'tak dianggap'  song tend to fall into the acousticness and danceability genres


In [ ]:
for i in groupby_songs_daily_rank_1.Songs.to_list():
    print('\nSONG : ', i.upper())
    print('-'*30)
    display(df_id[df_id['Songs']==i].sort_index().head(1)[['danceability','energy','acousticness','valence','popularity']])

# Explore songs that are in the popularity 100

- there is 1 song that is popular and contains the explicit word, namely greedy, and has been played 30 times
- there are 2 popular and non-explicit songs, namely Cruel Summer played 18 times, End of Beginning 2 times

In [ ]:
# songs that are popular and contain explicit words

true_explicit =df_id[(df_id.is_explicit==True) & (df_id.popularity==100)]
print('jumlah lagu yang populer dan eksplicit\n', true_explicit.Songs.value_counts(), '\n')
print(true_explicit[['snapshot_date','Songs']])


# popular and non-explicit songs

false_explicit = df_id[(df_id.is_explicit==False) & (df_id.popularity==100)]
print('\njumlah lagu yang populer dan non-eksplicit\n', false_explicit.Songs.value_counts())
print('\n',false_explicit[['snapshot_date','Songs']])

In [ ]:
# genre yang dimiliki oleh lagu yang mendapatkan kepopuleran 100 
genre_popularity_100 = df_id[df_id.popularity==100]
for song in genre_popularity_100['Songs'].unique():
    display(genre_popularity_100[genre_popularity_100['Songs'] == song].head(1)[['artists','Songs','danceability', 'energy', 'valence']])
    


.

On average, songs that are in the highest popularity (100) are more dominant towards energetic songs. Usually energetic tracks feel fast, loud and noisy. For example, death metal has high energy and valence, namely the positivity the music conveys, such as happiness or cheerfulness.

In [ ]:
# average popular song per genre
mean = genre_popularity_100.groupby('artists').agg(
{
    'danceability': 'mean',
    'energy' : 'mean',
    'valence' : 'mean',
    'acousticness' : 'mean'
})
print(mean)

mean.mean()


# correlation between the features

There are 3 features that appear to have the most correlation
- is_explicit - speechiness = 0.6 to 0.8
- energy - valence = 0.6 to 0.8
- energy - loudness = 0.8 to 1.0

In [ ]:
sns.heatmap(df_id.corr())

## Genre : danceability, energy and valence selama di indonesia

In the last few months there has been an increase in song playback with a level of danceability (mid of month 3 - end of month 4) followed by a decrease in energy level, there were 4 times when there was an increase in danceability and the highest value of danceability occurred on 11-4-2024:        


1. 2024-03-23 
    - danceability: 0.550720    
    - energy: 0.492220
    - valence: 0.386300
   

 
2. 2024-03-30  
    - danceability :0.550400 
    - energy: 0.491000    
    - valence :0.384040
           


3. 2024-04-09         
    - danceability: 0.551300   
    - energy: 0.484840
    - valence: 0.374340
           

4. 2024-04-11 
    - danceability : 0.553000
    - energy: 0.479780
    - valence: 0.371040

In [ ]:
time_series =df_id.groupby(['snapshot_date']).agg({
    'danceability': 'mean',
    'energy' : 'mean',
    'valence' : 'mean',
})
id_time_series = time_series.sort_index()
#print(id_time_series.head())

id_time_series.plot(
    y =['danceability', 'energy', 'valence'],
    color = ['blue', 'green', 'red'],
linestyle= '-', marker= 'o',
figsize =(30,20),
title = 'danceability, energy and valence over time in the indonesia',
xlabel = 'snapshotDate', ylabel='value')

plt.grid(True)
plt.legend(loc='upper right')
plt.show()

In [ ]:
# increase in playability of songs with danceability level (mid month 3 - end of month 4)
print(id_time_series.tail(60))